source/extensions/common/wasm/wasm_vm_base.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "extensions/common/wasm/wasm_vm.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+/**
+ * Wasm host stats.
+ */
+#define ALL_VM_STATS(COUNTER, GAUGE)                                                               \
+  COUNTER(created)                                                                                 \
+  COUNTER(cloned)                                                                                  \
+  GAUGE(active, NeverImport)
+
+struct VmStats {
+  ALL_VM_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+struct VmGlobalStats {
+  std::atomic<int64_t> active_;
+};
+
+// Wasm VM base instance. Provides common behavior (e.g. Stats).
+class WasmVmBase : public WasmVm {
+public:
+  WasmVmBase(Stats::ScopeSharedPtr scope, VmGlobalStats* global_stats_ptr,
+             absl::string_view runtime)
+      : scope_(scope), global_stats_ptr_(global_stats_ptr),
+        stats_(VmStats{
+            ALL_VM_STATS(POOL_COUNTER_PREFIX(*scope_, absl::StrCat("wasm_vm.", runtime, ".")),
+                         POOL_GAUGE_PREFIX(*scope_, absl::StrCat("wasm_vm.", runtime, ".")))}),
+        runtime_(std::string(runtime)) {
+    global_stats_ptr_->active_++;
+    stats_.created_.inc();
+    stats_.active_.set(global_stats_ptr_->active_);
+    ENVOY_LOG(debug, "WasmVm created {} now active", runtime_, global_stats_ptr_->active_);
+  }
+  virtual ~WasmVmBase() {
+    global_stats_ptr_->active_--;
```

Comment:
The old guidance for pre-increment and pre-decrement presumes a not very bright compiler.  That hasn't been the case for a while.  Similarly I have tested pass by value for std::shared_ptr vs std::move and in the cases where the compiler can easily prove that the std::shared_ptr is no longer live, it will elide the copy and do the move automatically.  IMO we should favor readability in such cases and trust but verify that our compiler is doing its job.

Reply:
FWIW the style guidelines are updated pretty often, and still (last I checked) specifically suggest pre-increment for class objects, even though it's possible for compilers to optimize this (as is the cast for atomics).

But I'm not worried about this, you can resolve & and leave as is, I'm just pointing out it violates style guide. I also don't personally find the post-increment form easier to read, but to each his own.

Reply:
Changed to pre-increment.